plutus-core/src/Language/PlutusCore/Constant/Meaning.hs

```diff
@@ -0,0 +1,266 @@
+a type class and generate a PLC kind at the Haskell's term level.
+-}
+
+type family GetArgs a :: [GHC.Type] where
+    GetArgs (a -> b) = a ': GetArgs b
+    GetArgs _        = '[]
+
+-- | A class that allows to derive a 'Monotype' for a builtin.
+class KnownMonotype term args res a | args res -> a, a -> res where
+    knownMonotype :: TypeScheme term args res
+
+-- | Once we've run out of term-level arguments, we return a 'TypeSchemeResult'.
+instance (res ~ res', KnownType term res) => KnownMonotype term '[] res res' where
+    knownMonotype = TypeSchemeResult Proxy
+
+-- | Every term-level argument becomes as 'TypeSchemeArrow'.
+instance (KnownType term arg, KnownMonotype term args res a) =>
+            KnownMonotype term (arg ': args) res (arg -> a) where
+    knownMonotype = Proxy `TypeSchemeArrow` knownMonotype
+
+-- | Delete all @x@s from a list.
+type family Delete x xs :: [a] where
+    Delete _ '[]       = '[]
+    Delete x (x ': xs) = Delete x xs
+    Delete x (y ': xs) = y ': Delete x xs
+
+-- | Delete all elements appearing in the first list from the second one and concatenate the lists.
+type family Merge xs ys :: [a] where
+    Merge '[]       ys = ys
+    Merge (x ': xs) ys = x ': Delete x (Merge xs ys)
+
+-- | Collect all unique variables (a variable consists of a textual name and a unique) in an @x@.
+type family ToBinds (x :: a) :: [(Symbol, Nat)]
+
+type instance ToBinds (TyVarRep name uniq) = '[ '(name, uniq) ]
+type instance ToBinds (TyAppRep fun arg) = TypeError ('Text "Not supported yet")
+
+-- | A standalone Haskell type either stands for a PLC type (when it's an 'Opaque') or
+-- it's an actual Haskell type, in which case it doesn't contain any PLC type variables.
+type family TypeToBinds (a :: GHC.Type) :: [(Symbol, Nat)] where
+    TypeToBinds (Opaque _ rep) = ToBinds rep
+    TypeToBinds _              = '[]
+
+type instance ToBinds (TypeScheme term '[]           res) = TypeToBinds res
+type instance ToBinds (TypeScheme term (arg ': args) res) =
+    Merge (TypeToBinds arg) (ToBinds (TypeScheme term args res))
+
+-- | A class that allows to derive a 'Polytype' for a builtin.
+class KnownPolytype (binds :: [(Symbol, Nat)]) term args res a | args res -> a, a -> res where
+    knownPolytype :: Proxy binds -> TypeScheme term args res
+
+-- | Once we've run out of type-level arguments, we start handling term-level ones.
+instance KnownMonotype term args res a => KnownPolytype '[] term args res a where
+    knownPolytype _ = knownMonotype
+
+-- | Every type-level argument becomes a 'TypeSchemeAll'.
+instance (KnownSymbol name, KnownNat uniq, KnownPolytype binds term args res a) =>
+            KnownPolytype ('(name, uniq) ': binds) term args res a where
+    knownPolytype _ = TypeSchemeAll @name @uniq Proxy (Type ()) $ \_ -> knownPolytype (Proxy @binds)
+
+-- See Note [Automatic derivation of type schemes]
+-- | Construct the meaning for a dynamic built-in function by automatically deriving its
```

Comment:
> Does this now mean one which uses the "dynamic" input to the semantics?

Yes. So... I mean, we kinda just used the term wrongly: why was `charToString` a dynamic builtin if both `Char` and `String` were in the default universe anyway? We should've made it static. And `Append` as well. And then we're only left with `Trace`. And now the two meanings of this dichotomy coincide.

If you have a better terminology, please suggest it, I didn't feel brave enough to dump what I thought was rather precise terms just because we used them previously in a slightly different manner and not for good reasons.

Reply:
So the difference is that before the *set* of builtins could be changed at runtime, whereas now the *semantics* of the builtins can change at runtime. So yes, there's something dynamic about it, but it's quite different to how it was before.

I don't have a better suggestion right now, but I'll think about it.

Reply:
> it's quite different to how it was before.

We can pretend the before has never happened.